### DISCUSSION TITLE
Adios2 with Melissa

### DISCUSSION THREAD

**Discussion by mschouler** [ornladios/ADIOS2#3675] [Q&A]
Hi, I am a developer in the [Melissa](https://melissa.gitlabpages.inria.fr/melissa/) team and we are currently investigating adios2 as a way to improve performances of our software.

In order to give you a bit of context, Melissa is a framework that coordinates ensemble runs to perform large scale and online sensitivity analysis or deep surrogate training. Our usual workflow consists of multiple data producers (any MPI  code instrumented with our API) which dynamically connect to a parallel Python server and send this server data as they are produced (either in an MxN or a round robin fashion). Our point is to avoid files by directly communicating data through our API based on ZeroMQ sockets.

The main inconvenient in our current configuration is that ZeroMQ does not enable to take full advantage of the data transport layer on supercomputers which is why adios2 seemed like a good alternative. In substituting our current API with adios2, we would still like to handle client connection and manage message reception somehow asynchronously on the server side.

So far we have managed to build a preliminary working version with adios2. However, it is pretty sequential in the sense that as a data producer is detected as running by the server rank 0, its id is broadcasted to all ranks and the corresponding engine is opened on all ranks at the same time. Regarding message extraction, we are basically looping over all active engines trying to extract data if they are holding any.  

From the adios2 documentation, it seems that engine (SST in our case) Open/Close are collective functions (not sure about BeginStep/EndStep). It also says that adios2 should be considered as not thread-safe. Indeed, as soon as we are trying to optimize our approach with threads to handle engine opening for instance, the server runs into deadlocks (same with message extraction).

Given our preliminary tests, we are hence wondering if adios2 is actually well suited for our purpose or not. If so, what do you think would be the optimal way to handle client connection and message reception?

Best regards,
Marc

**Comment by eisenhauer** [comment]
I'm going to suggest that @anagainaru chime in because she has been working with some of these sorts of issues, but in the meantime, I can address the SST questions.  Yes, in general Open is a collective operation across all the engines.  ADIOS' focus is on optimizing collective IO and in some sense the sweet spot is supporting applications like parallel timestepped simulations where all ranks dump data at the same time and in the aggregate it forms a global data structure.  So while BeginStep might or might not be collective, EndStep pretty much always is.  That said, @anagainaru has been using SST in more of a job-distribution sort of way by having multiple readers connect to a single producer and using SST's "OnDemand" value for the StepDistributionMode parameter.  Hopefully she can fill you in on the details.

**Reply by mschouler** [reply]
Thank you very much for your reply!

**Comment by anagainaru** [comment]
Yes, @eisenhauer is right, I have been looking at this exact scenario. 

Typically, ADIOS was developed for simulations that generate data in steps. When you connect a producer with a consumer, each time a producer generates a step (which means all the puts between a `BeginStep` and `EndStep`) this step is streamed to all the consumers (for the default behavior). You are the one defining what a step looks like, for example for cancer research where multiple MRIs are used to train a classification model, each step is represented by a batch of MRIs. I would recommend defining the steps based on the data the producers are waiting to receive (the more data you include in one step the more work for the consumers but you might have to wait longer until the producer sends anything). Only once an `EndStep` call is executed by the producer, the step is streamed to consumers. We currently have 3 ways of doing this: all to all (all consumers receive all steps),  round robin (consumers receive steps in round robin order) and on demand (consumers ping the producers when they need data and they receive the next available step). For AI applications where consumers train surrogate models I use the on demand strategy. The streaming looks sequential if you have one consumer but it will be concurrent for your MxN example.

I added a simple example for using on demand using SST here: https://github.com/anagainaru/ADIOS2-addons/tree/main/DataStreaming/ADIOS_ondemand

The `adios.xml` file is used in this example to specify the number of consumers and the streaming strategy (in this case 2 consumers and on demand):

```
        <parameter key="RendezvousReaderCount" value="2"/>
        <parameter key="StepDistributionMode" value="OnDemand"/>
```

To run the code:
```
./sstWriter 100 2 & ./sstReader 100 2 & ./sstReader 100 2
```

Let me know if this makes sense. I can take a look at code snippets if you'd like me to figure out why your testcase looks sequential.

**Reply by mschouler** [reply]
Thank you very much for this detailed answer and the code example, it makes total sense.

I am sorry for the confusion, it seems that my use of the term _sequential_ was misleading. Rephrasing my original post, our preliminary tests showed that for a parallel data consumer (we call it server) using adios2 (SST engine and "AllToAll" step distribution mode): 
1. The server ranks need to stay strongly synchronized when opening engines (all ranks need to open the same engine at the same time). For this reason, it does not seem possible to use a thread pool where each thread is associated to an engine and tries to call `Open` for this engine.
2. In the same way, it does not seem possible to use a per engine thread pool to call `BeginStep/Get/EndStep` on its associated engine in order to automatically extract time-steps as they become available. The easy but inefficient solution I came up with was to loop through all engines and extract data when available. Again, this seems to work because ranks are synchronized when a time-step is extracted from an engine. 

In comparison, the original Melissa implementation uses sockets which are checked in a [`run`](https://gitlab.inria.fr/melissa/melissa/-/blob/develop/melissa/server/base_server.py#L405-444) function. The main point being that each rank can thus loop asynchronously over its own series of sockets and deal with data processing when appropriate.

With the "OnDemand" step distribution mode, the data consumers have control of when to ask for data but not of which rank gets which steps, is that correct? Since the data availability strongly depends on the data producer speed, I guess the consumer `EndStep` call would be blocking right? Could this somehow be handled in a separate thread?

**Reply by anagainaru** [reply]
Could you share a minimal piece of code that will allow us to reproduce what you are describing? 
If each thread uses it's own adios object, we expect there will be no need for synchronized opens. The consumer will wait until all expected consumers call open before starting the streaming (in my example `RendezvousReaderCount` is set to 2, so the producer will block until 2 consumers call open). 

> With the "OnDemand" step distribution mode, the data consumers have control of when to ask for data but not of which rank gets which steps, is that correct?

That is correct, when the consumer asks for data it will receive whatever step is firstly available.

> Could this somehow be handled in a separate thread?

In principle this should work but again I would have to look at some code to make sure.

**Reply by mschouler** [reply]
Hi @anagainaru, thank you for your answer.

You will find in [this repository](https://gitlab.inria.fr/mschoule/adios2-melissa-simple-demo), a series of scripts I quickly put together to reproduce a Melissa like structure. I couldn't make it shorter though, sorry. The server script shows how we would like to dynamically handle **engine opening** and **data processing** with threads.

It is probably not the best solution but the **data processing** seems to work when it is handled in a separate thread (which is pretty encouraging). I guess there is no easy way to know which engines are holding data without beginning a step right? Regarding the dynamic **engine opening**, it does not work in a thread pool even with a single sequential client and a sequential server. Maybe I'm missing something or even more likely I am doing something wrong but I cannot figure out what..

As discussed in the last section of the [repo. README](https://gitlab.inria.fr/mschoule/adios2-melissa-simple-demo#simplification-wrt-real-melissa-conditions), it is really important for Melissa to be able to handle client connection dynamically since new clients are susceptible to connect at any time during the study. Opening engines before starting the message reception is therefore not a realistic solution. 

Finally, the sensitivity analysis major constraints is that all server ranks need to access their portion of data for all produced steps. However, your example of the "OnDemand" data distribution mode shows that there could be significant variations in the number of received steps per reader. I am afraid that could be a problem in our case of figure.

**edit**: our institutional gitlab is down until Monday 11 am (time of Paris).

**Reply by anagainaru** [reply]
@mschouler thanks for sharing the code. I will take a look on Monday when the gitlab is up.

**Reply by mschouler** [reply]
Hi @anagainaru, I had a bit of time and fixed a stupid error of my doing in the pool thread submission command. 

The `--thread_open` option now works for a sequential server and multiple clients or for a parallel server and a single client. However it fails for a parallel server and multiple clients. I feel like it could be due to the server ranks trying to open different engines at the same time.

**Reply by anagainaru** [reply]
Hey @mschouler I was able to reproduce your problem using your code. I will try to debug it this week and come back to you.

**Reply by anagainaru** [reply]
OK, I think I understand what you need from adios now. You basically have multiple clients (running on multiple ranks) generating data that needs to be processes by one server (running on multiple ranks). To do this, yes you need multiple reader engines on the server side for each of the clients. Once you do that, you can periodically poll each of those engines with a BeginStep with zero timeout to see if they’ve got data ready for you.

So we believe the issue with your code is (as you guessed) the Opens when using multiple threads. Getting the ops for different Open’s intertwined on a single communicator is an issue that might cause the Open to fail. So my first question is, do you need to create threads for Opens? Would sequential Opens kill your performance? As long as all ranks call Opens on engines in the same order you will be ok. 

A possible solution (that we did not test for but that I could look into if you really need to parallelize Opens) is to use different communicators in different threads for the Open.

On the read side, things are much nicer. Right now you are locking for receiving data, but you could create separate threads to read from each engine and merge the results somehow through a thread-safe queue.

**Reply by mschouler** [reply]
Thank you very much @anagainaru. This gives us a lot to think. 

We will certainly do some testing at scale in order to assess to what extent Opens management affects performances. Using distinct communicators sounds like a really interesting lead, we will probably test it as well.

Overall, it seems like adios2 is very well suited for our needs. We will let you know how it goes as we progress in its integration with Melissa.

**Reply by anagainaru** [reply]
Sounds good, keep us posted
